src/com/amazon/ion/benchmark/ReadGeneralConstraints.java

```diff
@@ -1,52 +1,72 @@
 
     /**
-     * Get general constraints of Ion Schema and call the relevant generator method based on the type.
+     * Getting the constructed data which is conformed with ISL and writing data to the output file.
      * @param size is the size of the output file.
-     * @param path is the path of the Ion Schema file.
+     * @param schema an Ion Schema loaded by ion-schema-kotlin.
      * @param format is the format of the generated file, select from set (ion_text | ion_binary).
      * @param outputFile is the path of the generated file.
-     * @throws Exception if errors occur when reading and writing data.
+     * @throws Exception if errors occur when writing data.
      */
-    public static void readIonSchemaAndGenerate(int size, String path, String format, String outputFile) throws Exception {
-        try (IonReader reader = IonReaderBuilder.standard().build(new BufferedInputStream(new FileInputStream(path)))) {
-            IonDatagram schema = LOADER.load(reader);
-            for (int i = 0; i < schema.size(); i++) {
-                IonValue schemaValue = schema.get(i);
-                // Assume there's only one constraint between schema_header and schema_footer, if more constraints added, here is the point where developers should start.
-                if (schemaValue.getType().equals(IonType.STRUCT) && schemaValue.getTypeAnnotations()[0].equals(IonSchemaUtilities.KEYWORD_TYPE)) {
-                    IonStruct constraintStruct = (IonStruct) schemaValue;
-                    //Construct the writer and pass the constraints to the following writing data to files process.
-                    File file = new File(outputFile);
-                    try (IonWriter writer = WriteRandomIonValues.formatWriter(format, file)) {
-                        WriteRandomIonValues.writeRequestedSizeFile(size, writer, file, constraintStruct);
-                    }
-                    // Print the successfully generated data notification which includes the file path information.
-                    WriteRandomIonValues.printInfo(outputFile);
-                }
+    public static void constructAndWriteIonData(int size, Schema schema, String format, String outputFile) throws Exception {
+        // Assume there's only one type definition between schema_header and schema_footer.
+        // If more constraints added, here is the point where developers should start.
+        Type schemaType = schema.getTypes().next();
+        ReparsedType parsedTypeDefinition = new ReparsedType(schemaType);
+        CountingOutputStream outputStreamCounter = new CountingOutputStream(new FileOutputStream(outputFile));
+        long count = 0;
+        try (IonWriter writer = formatWriter(format, outputStreamCounter)) {
+            while (count <= size) {
+                IonValue constructedData = DataConstructor.constructIonData(parsedTypeDefinition);
+                constructedData.writeTo(writer);
+                count = outputStreamCounter.getCount();
```

Comment:
> In order for the count to update, the writer needs to be flushed. We may want to do this after every X values are written.

Thanks for pointing this out. Yes, we need to flush the buffered stream in writer to make sure the size of written data is up to date. 
And here is another thought I have:
The motivation of using counter to count OutputStream is to avoid the original while loop logic and checking the OutputStream size instead of checking the output file size. 

Original logic:
```
 public static void writeRequestedSizeFile(int size, IonWriter writer, File file, IonStruct constraintStruct) throws Exception {
        int currentSize = 0;
        int count = 0;
        // Determine how many values should be written before the writer.flush(), and this process aims to reduce the execution time of writer.flush().
        while (currentSize <= 0.05 * size) {
            WriteRandomIonValues.writeDataToFile(writer, constraintStruct);
            count += 1;
            writer.flush();
            currentSize = (int) file.length();
        }
        while (currentSize <= size) {
            for (int i = 0; i < count; i++) {
                WriteRandomIonValues.writeDataToFile(writer, constraintStruct);
            }
            writer.flush();
            currentSize = (int) file.length();
        }
    }
```
We do need to flush the writer, but flush after X written data would take the method back to the logic we were trying to avoid. Probably the further consideration/discussion is needed here to come up with a more straightforward way to monitor the size of written data.

Reply:
The main thing that is bad about the old logic was repeatedly calling `File.length()`, which has to go all the way to the file system each call. `CountingOutputStream` is at least an improvement over that.

Reply:
> The main thing that is bad about the old logic was repeatedly calling `File.length()`, which has to go all the way to the file system each call. `CountingOutputStream` is at least an improvement over that.

Yes, for now it would be fine to keep the logic of flushing after X written data then using `CoutingOutputStream` as counter to monitor the size. The updates will be included in the following commit.